amethyst_ui/src/drag.rs

```diff
@@ -0,0 +1,157 @@
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, HashSet};
+use std::marker::PhantomData;
+
+use amethyst_core::{
+    ecs::{
+        Component, DenseVecStorage, Entities, Entity, Join, Read, ReadExpect, ReadStorage,
+        ReaderId, System, SystemData, Write, WriteStorage,
+    },
+    math::Vector2,
+    shrev::EventChannel,
+    Hidden, HiddenPropagate,
+};
+use amethyst_derive::SystemDesc;
+use amethyst_input::{BindingTypes, InputHandler};
+use amethyst_window::ScreenDimensions;
+
+use crate::{targeted_below, Interactable, UiEvent, UiEventType, UiTransform};
+
+/// Component that denotes whether a given ui widget is draggable.
+/// Requires UiTransform to work, and its expected way of usage is
+/// through UiTransformData prefab.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Draggable;
+
+impl Component for Draggable {
+    type Storage = DenseVecStorage<Self>;
+}
+
+#[derive(Debug, SystemDesc)]
+#[system_desc(name(DragWidgetSystemDesc))]
+pub struct DragWidgetSystem<T: BindingTypes> {
+    #[system_desc(event_channel_reader)]
+    ui_reader_id: ReaderId<UiEvent>,
+
+    /// hashmap whose keys are every entities being dragged,
+    /// and whose element is a tuple whose first element is
+    /// the original mouse position when drag first started,
+    /// and second element the mouse position one frame ago
+    #[system_desc(skip)]
+    record: HashMap<Entity, (Vector2<f32>, Vector2<f32>)>,
+
+    phantom: PhantomData<T>,
+}
+
+impl<T> DragWidgetSystem<T>
+where
+    T: BindingTypes,
+{
+    pub fn new(ui_reader_id: ReaderId<UiEvent>) -> Self {
+        Self {
+            ui_reader_id,
+            record: HashMap::new(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'s, T> System<'s> for DragWidgetSystem<T>
+where
+    T: BindingTypes,
+{
+    type SystemData = (
+        Entities<'s>,
+        Read<'s, InputHandler<T>>,
+        ReadExpect<'s, ScreenDimensions>,
+        ReadStorage<'s, Hidden>,
+        ReadStorage<'s, HiddenPropagate>,
+        ReadStorage<'s, Draggable>,
+        ReadStorage<'s, Interactable>,
+        Write<'s, EventChannel<UiEvent>>,
+        WriteStorage<'s, UiTransform>,
+    );
+
+    fn run(
+        &mut self,
+        (
+            entities,
+            input_handler,
+            screen_dimensions,
+            hiddens,
+            hidden_props,
+            draggables,
+            interactables,
+            mut ui_events,
+            mut ui_transforms,
+        ): Self::SystemData,
+    ) {
+        if let Some((mouse_x, mouse_y)) = input_handler.mouse_position() {
+            let mouse_pos = Vector2::new(mouse_x, screen_dimensions.height() - mouse_y);
+
+            let mut click_stopped: HashSet<Entity> = HashSet::new();
+
+            for event in ui_events.read(&mut self.ui_reader_id) {
```

Comment:
Hm, something tells me the `ui_events.read(&mut self.ui_reader_id)` should be outside the `if let Some((mouse_x, mouse_y))`, so that, if there are events, but no mouse position, the event channel doesn't get clogged up by this reader not reading them.

I'm not sure when it's possible for the mouse position to be `None` though.

That said, the `ClickStart` and `ClickStop` events need to be collected (because it would conflict with the `ui_events.single_write` a bit further down).

Reply:
That is a fair point. Thanks for pointing that out. Will work on it right away.